### PR TITLE
Java implementation

### PR DESCRIPTION
**Pull request by azanin**
Java implementation to produce & verify TrueLayer API requests signatures.

Java usage:
```java
// `Tl-Signature` value to send with the request.
Signer.from(kid, privateKey)
        .header("Idempotency-Key", idempotencyKey)
        .method("post")
        .path(path)
        .body(body)
        .sign();
```

### CODE REVIEW

File: java/lib/src/main/java/truelayer/signing/Signer.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+/**
+ * Builder to generate a Tl-Signature header value using a private key.
+ */
+final public class Signer {
+
+    private final String kid;
+    private final ECPrivateKey ecPrivateKey;
+
+    private String path = "";
+    private String method = "";
```
Comment (tl-marco-tormento) [inline java/lib/src/main/java/truelayer/signing/Signer.java:25]: Method should have a default value (e.g. POST).

File: java/lib/src/main/java/truelayer/signing/Verifier.java [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+import java.util.stream.Stream;
+
+/**
+ * Builder to verify a request against a `Tl-Signature` header using a public key.
+ */
+final public class Verifier {
+
+    private final ECPublicKey publicKey;
+
+    private String method = "";
```
Comment (tl-marco-tormento) [inline java/lib/src/main/java/truelayer/signing/Verifier.java:22]: Method should have a default value (e.g. POST).

File: java/lib/src/test/java/truelayer/signing/UsageTest.java [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+                .path(path)
+                .header("X-Whatever", "aoitbeh")
+                .header("Idempotency-Key", idempotencyKey)
+                .body(body)
+                .requiredHeader("Idempotency-Key")
+                .verify(tlSignature); // should not throw
+    }
+
+    @Test
+    public void verifyBodyStaticSignature() throws IOException {
```
Comment (alexheretic) [inline java/lib/src/test/java/truelayer/signing/UsageTest.java:45]: The "body static" probably comes from the rust lib that supports body-only signing too.

```suggestion
    public void verifyStaticSignature() throws IOException {
```

File: java/lib/src/test/java/truelayer/signing/UsageTest.java [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+                .path(path)
+                .header("X-Whatever-2", "t2345d")
+                .header("Idempotency-Key", idempotencyKey)
+                .body(body)
+                .verify(tlSignature); // should not throw
+    }
+
+
+    @Test
+    public void fullRequestSignatureMethodMismatch() throws IOException {
```
Comment (alexheretic) [inline java/lib/src/test/java/truelayer/signing/UsageTest.java:65]: We can probably remove the `fullRequest` prefix as java won't support v1/body-only signing.

**Review by alexheretic — APPROVED** [review]

File: java/lib/src/main/java/truelayer/signing/Verifier.java [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+
+    private String path = "POST";
```
Comment (alexheretic) [inline java/lib/src/main/java/truelayer/signing/Verifier.java:23]: wait not this :laughing:

File: java/lib/src/main/java/truelayer/signing/Verifier.java (hunk 2 above)
Comment (alexheretic) [reply java/lib/src/main/java/truelayer/signing/Verifier.java:21]: Actually I don't think _verifier_ method should have a default, just _sign_.

File: java/lib/src/main/java/truelayer/signing/Verifier.java (hunk 5 above)
Comment (azanin) [reply java/lib/src/main/java/truelayer/signing/Verifier.java:23]: oh :/ stupid me

**Comment by alexheretic** [comment]
Should update the main README too & add an initial CHANGELOG.md like the other langs.

**Review by alexheretic — APPROVED** [review]
:medal_sports:

**Pull request merged**